Typechecking assertion panics on pulumi convert

### What happened?

Using v3.191.0 and the PCL below, I run `pulumi convert --from pcl --language typescript`

and am rewarded with the following panic:

`panic: fatal: An assertion has failed: error typechecking expression: debug.pp:22,45-57: unknown property 'dns_name' among [dnsName hostedZoneId];`

Gist of stack trace: https://gist.github.com/guineveresaenger/58b7f15d158ea4a3b005ed4abce82ebe


### Example

Pulumi PCL
```
resource "ptfeService" "aws:ec2/vpcEndpoint:VpcEndpoint" {
  __logicalName     = "ptfe_service"
  vpcId             = vpcId
  serviceName       = ptfeServiceConfig
  vpcEndpointType   = "Interface"
  securityGroupIds  = [ptfeServiceAwsSecurityGroup.id]
  subnetIds         = [subnetIds]
  privateDnsEnabled = false
}

internal = invoke("aws:route53/getZone:getZone", {
  name        = "vpc.internal."
  privateZone = true
  vpcId       = vpcId
})

resource "ptfeServiceRecord" "aws:route53/record:Record" {
  __logicalName = "ptfe_service"
  zoneId        = internal.zoneId
  name          = "ptfe.${internal.name}"
  type          = "CNAME"
  ttl           = "300"
  records       = [ptfeService.dnsEntries[0]["dns_name"]]
}
```

### Output of `pulumi about`

```
CLI          
Version      3.191.0
Go Version   go1.25.0
Go Compiler  gc

Host     
OS       darwin
Version  15.5
Arch     arm64

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/guinevere-pulumi-corp
User           guinevere-pulumi-corp
Organizations  guinevere-pulumi-corp, pulumi
Token type     personal
```

```
pulumi plugin ls
NAME        KIND       VERSION  SIZE    INSTALLED     LAST USED
aws         resource   7.6.0    832 MB  23 hours ago  23 hours ago
cloudflare  resource   6.6.0    168 MB  19 hours ago  19 hours ago
gcp         resource   8.41.1   200 MB  20 hours ago  20 hours ago
gcp         resource   7.38.0   205 MB  20 hours ago  20 hours ago
terraform   converter  1.2.3    42 MB   2 days ago    2 days ago
```

### Additional context

This was discovered attempting to upgrade the pulumi-terraform-bridge. Pulumi-aws will fail to build its schema.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

Reopening to track the new failure:

```
guin pulumi/convert-debug🦉 pulumi convert --from pcl --language typescript debug.pp
Converting from pcl...
Converting to nodejs...
panic: fatal: An assertion has failed: unexpected type error: debug.pp:22,45-57: unknown property 'dns_name' among [dnsName hostedZoneId]; 

goroutine 22 [running]:

github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.failfast(...)

	/private/tmp/pulumi-20250912-4901-gujqwr/sdk/go/common/util/contract/failfast.go:23

github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.Assertf(0x0?, {0x102d7b262?, 0x1?}, {0x14002384ef8?, 0x1030a85a0?, 0x10316cca0?})

	/private/tmp/pulumi-20250912-4901-gujqwr/sdk/go/common/util/contract/assert.go:35 +0xd4

github.com/pulumi/pulumi/pkg/v3/codegen/nodejs.(*generator).parseProxyApply(0x14006823630, 0x14002385098, {0x1400d92c060?, 0x0?, 0x14002385008?}, {0x10325ce00, 0x14003c441e0})

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/nodejs/gen_program_lower.go:130 +0x744

github.com/pulumi/pulumi/pkg/v3/codegen/nodejs.(*generator).lowerProxyApplies.func1({0x10325d300, 0x14001a38f00?})

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/nodejs/gen_program_lower.go:228 +0x18c

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.visitFunctionCall(0x14001a38f00, 0x14001a38f00?, 0x14002385878)

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/hcl2/model/visitor.go:157 +0x108

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.VisitExpression({0x10325d300?, 0x14001a38f00?}, 0x14002385268?, 0x14002385878)

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/hcl2/model/visitor.go:323 +0x2b4

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.visitExpressions({0x1400682d370, 0x1, 0x1}, 0x10323b0f8, 0x14002385878)

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/hcl2/model/visitor.go:273 +0xb4

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.visitFunctionCall(0x14001a38f80, 0x14001a38f80?, 0x14002385878)

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/hcl2/model/visitor.go:152 +0x44

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.VisitExpression({0x10325d300?, 0x14001a38f80?}, 0x0?, 0x14002385878)

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/hcl2/model/visitor.go:323 +0x2b4

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.visitExpressions({0x1400682cc00, 0x1, 0x1}, 0x10323b0f8, 0x14002385878)

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/hcl2/model/visitor.go:273 +0xb4

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.visitTupleCons(0x14000378540, 0x14000378540?, 0x14002385878)

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/hcl2/model/visitor.go:247 +0x44

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.VisitExpression({0x10325d980?, 0x14000378540?}, 0x14002385688?, 0x14002385878)

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/hcl2/model/visitor.go:341 +0xe0

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.visitExpressions({0x1400682d390, 0x1, 0x1}, 0x10323b0f8, 0x14002385878)

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/hcl2/model/visitor.go:273 +0xb4

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.visitFunctionCall(0x14001a39000, 0x14001a39000?, 0x14002385878)

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/hcl2/model/visitor.go:152 +0x44

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.VisitExpression({0x10325d300?, 0x14001a39000?}, 0x2ac?, 0x14002385878)

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/hcl2/model/visitor.go:323 +0x2b4

github.com/pulumi/pulumi/pkg/v3/codegen/nodejs.(*generator).lowerProxyApplies(0x10325d980?, {0x10325d300?, 0x14001a39000?})

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/nodejs/gen_program_lower.go:239 +0x44

github.com/pulumi/pulumi/pkg/v3/codegen/nodejs.(*generator).lowerExpression(0x14006823630, {0x10325d980?, 0x14000378540?}, {0x10325cf00, 0x140068227d0})

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/nodejs/gen_program_expressions.go:57 +0x15c

github.com/pulumi/pulumi/pkg/v3/codegen/nodejs.(*generator).genResourceDeclaration.(*generator).genResourceDeclaration.func1.func9()

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/nodejs/gen_program.go:968 +0x458

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model/format.(*Formatter).Indented(...)

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/hcl2/model/format/gen.go:85

github.com/pulumi/pulumi/pkg/v3/codegen/nodejs.(*generator).genResourceDeclaration.func1(...)

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/nodejs/gen_program.go:945

github.com/pulumi/pulumi/pkg/v3/codegen/nodejs.(*generator).genResourceDeclaration(0x14006823630, {0x1032451a8, 0x140068376e0}, 0x140001f5080, 0x1)

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/nodejs/gen_program.go:1118 +0x171c

github.com/pulumi/pulumi/pkg/v3/codegen/nodejs.(*generator).genResource(...)

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/nodejs/gen_program.go:1126

github.com/pulumi/pulumi/pkg/v3/codegen/nodejs.(*generator).genNode(0x140023867b8?, {0x1032451a8?, 0x140068376e0?}, {0x10325bcb8?, 0x140001f5080?})

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/nodejs/gen_program.go:790 +0x64

github.com/pulumi/pulumi/pkg/v3/codegen/nodejs.GenerateProgram.func2()

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/nodejs/gen_program.go:121 +0x58

github.com/pulumi/pulumi/pkg/v3/codegen/nodejs.GenerateProgram.func1(0x1030bbe80?)

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/nodejs/gen_program.go:113 +0x24

github.com/pulumi/pulumi/pkg/v3/codegen/nodejs.GenerateProgram(0x14000378680)

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/nodejs/gen_program.go:119 +0x64c

github.com/pulumi/pulumi/pkg/v3/codegen/nodejs.GenerateProject({0x1400041a200, 0x32}, {{0x1400037d5c0, 0xd}, {{0x0, 0x0}, 0x0}, {0x0, 0x0}, 0x0, ...}, ...)

	/private/tmp/pulumi-20250912-4901-gujqwr/pkg/codegen/nodejs/gen_program.go:175 +0x3c

main.(*nodeLanguageHost).GenerateProject(0x14000260100, {0x103200aa0?, 0x40?}, 0x140000f1080)

	/private/tmp/pulumi-20250912-4901-gujqwr/sdk/nodejs/cmd/pulumi-language-nodejs/main.go:1617 +0x38c

github.com/pulumi/pulumi/sdk/v3/proto/go._LanguageRuntime_GenerateProject_Handler.func1({0x103252b40?, 0x1400033a5a0?}, {0x1031c0ea0?, 0x140000f1080?})

	/private/tmp/pulumi-20250912-4901-gujqwr/sdk/proto/go/language_grpc.pb.go:688 +0xcc

github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1({0x103252b40, 0x14000263ec0}, {0x1031c0ea0, 0x140000f1080}, 0x1400026c5e0, 0x14000402510)

	/Users/brew/Library/Caches/Homebrew/go_mod_cache/pkg/mod/github.com/grpc-ecosystem/grpc-opentracing@v0.0.0-20180507213350-8e809c8a8645/go/otgrpc/server.go:57 +0x2a4

github.com/pulumi/pulumi/sdk/v3/proto/go._LanguageRuntime_GenerateProject_Handler({0x103200aa0, 0x14000260100}, {0x103252b40, 0x14000263ec0}, 0x140000f1000, 0x14000326000)

	/private/tmp/pulumi-20250912-4901-gujqwr/sdk/proto/go/language_grpc.pb.go:690 +0x140

google.golang.org/grpc.(*Server).processUnaryRPC(0x14000332200, {0x103252b40, 0x1400023c3c0}, 0x1400018e360, 0x140002626c0, 0x103b7ac18, 0x0)

	/Users/brew/Library/Caches/Homebrew/go_mod_cache/pkg/mod/google.golang.org/grpc@v1.72.1/server.go:1405 +0xbd0

google.golang.org/grpc.(*Server).handleStream(0x14000332200, {0x103253090, 0x140006c84e0}, 0x1400018e360)

	/Users/brew/Library/Caches/Homebrew/go_mod_cache/pkg/mod/google.golang.org/grpc@v1.72.1/server.go:1815 +0x858

google.golang.org/grpc.(*Server).serveStreams.func2.1()

	/Users/brew/Library/Caches/Homebrew/go_mod_cache/pkg/mod/google.golang.org/grpc@v1.72.1/server.go:1035 +0x74

created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 68

	/Users/brew/Library/Caches/Homebrew/go_mod_cache/pkg/mod/google.golang.org/grpc@v1.72.1/server.go:1046 +0x120

error: could not generate output program: error reading from server: EOF
```

--- a/changelog/pending/20250925--programgen--fix-panic-in-parseproxyapply-when-encoutering-warning-diagnostics.yaml
+++ b/changelog/pending/20250925--programgen--fix-panic-in-parseproxyapply-when-encoutering-warning-diagnostics.yaml
@@ -1,0 +1,4 @@
+changes:
+- type: fix
+  scope: programgen
+  description: Fix panic in parseProxyApply when encoutering warning diagnostics

--- a/pkg/codegen/nodejs/gen_program_lower.go
+++ b/pkg/codegen/nodejs/gen_program_lower.go
@@ -127,7 +127,7 @@ func (g *generator) parseProxyApply(parameters codegen.Set, args []model.Express
 	}
 
 	diags := arg.Typecheck(false)
-	contract.Assertf(len(diags) == 0, "unexpected type error: %v", diags)
+	contract.Assertf(!diags.HasErrors(), "unexpected type error: %v", diags)
 	return arg, true
 }
 

--- a/pkg/codegen/pcl/binder_resource.go
+++ b/pkg/codegen/pcl/binder_resource.go
@@ -572,7 +572,11 @@ func (b *binder) bindResourceBody(node *Resource) hcl.Diagnostics {
 						}),
 					}
 					diags := condExpr.Typecheck(false)
-					contract.Assertf(len(diags) == 0, "failed to typecheck conditional expression: %v", diags)
+					checkDiagnostics := len(diags) == 0
+					if b.options.skipResourceTypecheck {
+						checkDiagnostics = !diags.HasErrors()
+					}
+					contract.Assertf(checkDiagnostics, "failed to typecheck conditional expression: %v", diags)
 
 					node.VariableType = condExpr.Type()
 				case model.InputType(model.NumberType).ConversionFrom(typ) == model.SafeConversion:
@@ -593,7 +597,11 @@ func (b *binder) bindResourceBody(node *Resource) hcl.Diagnostics {
 						Value: model.VariableReference(resourceVar),
 					}
 					diags := rangeExpr.Typecheck(false)
-					contract.Assertf(len(diags) == 0, "failed to typecheck range expression: %v", diags)
+					checkDiagnostics := len(diags) == 0
+					if b.options.skipResourceTypecheck {
+						checkDiagnostics = !diags.HasErrors()
+					}
+					contract.Assertf(checkDiagnostics, "failed to typecheck range expression: %v", diags)
 
 					rangeValue = model.IntType
 

--- a/pkg/codegen/python/gen_program_lower.go
+++ b/pkg/codegen/python/gen_program_lower.go
@@ -72,7 +72,7 @@ func (g *generator) parseProxyApply(parameters codegen.Set, args []model.Express
 	}
 
 	diags := arg.Typecheck(false)
-	contract.Assertf(len(diags) == 0, "unexpected diagnostics: %v", diags)
+	contract.Assertf(!diags.HasErrors(), "unexpected error diagnostics: %v", diags)
 	return arg, true
 }
 

